### PR TITLE
[Enhancement]: Scala Reflection: get default value for constructor parameters

### PR DESCRIPTION
**Pull request by LeicongLi**
## What changes were proposed in this pull request?
There would be an error if we try to load a module from existing protobuf file while some modules get updated with new constructing parameters.
To solve this, we include the functionality to get parameter's default value if this parameter has not been saved in protobuf model. 

## How was this patch tested?
To be tested

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializable.scala [hunk 1]
```diff
@@ -121,26 +121,49 @@ trait ModuleSerializable extends Loadable with Savable{
...
+          val attrValue = modelAttributes.get(pname)
+          DataConverter.getAttributeValue(context, attrValue)
+        } else { // parameter not found, get its default value
+          getPrimCtorDefaultParamValue(instanceMirror, param, i)
         }
-        i += 1
-      })
+        args(i) = pvalue
+      }
+      println(i, args(i))
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializable.scala]: remove this line

Comment (LeicongLi) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializable.scala]: removed

**Comment by LeicongLi** [comment]
jenkins result: http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/580/

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializable.scala [hunk 2]
```diff
@@ -121,26 +121,48 @@ trait ModuleSerializable extends Loadable with Savable{
...
-          val value = DataConverter.getAttributeValue(context, attribute)
-          args(i) = value
+
+    val clsMirror = universe.runtimeMirror(cls.getClassLoader)
+    val clsSymbol = clsMirror.classSymbol(cls)
+
+    /*
+    https://www.scala-lang.org/api/2.10.7/#scala.reflect.api.Symbols$Symbol
+    this line tries to get companion object of the class;
+    through the companion, default values can be access by calling
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializable.scala]: accessed

**Comment by wzhongyuan** [comment]
Have your tested with both 2.10 and 2.11 ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializable.scala (hunk 2 above)
Comment (LeicongLi) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializable.scala]: fixed typo

**Comment by LeicongLi** [comment]
> Have your tested with both 2.10 and 2.11 ?

Yes.
Tested using both 2.10 and 2.11 to load existing trained models, such as ResNet, DenseNet, MobileNet, etc.

**Pull request merged**